### PR TITLE
feat: basic logic converting csv into db (MAPCO-3357)

### PR DESCRIPTION
**Pull request by ronenkapelian**
read provided csv file with parts data, addapting + converting + uploading to related polygonParts DB

### CODE REVIEW

File: config/local.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (asafMasa) [inline config/local.json]: git ignore local file

Comment (asafMasa) [reply config/local.json]: revert

File: package.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "polygon-parts-cli",
+  "version": "1.0.0",
+  "description": "",
```
Comment (asafMasa) [inline package.json]: add description and readme

File: src/dataSource.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+class FileToDB {
+  constructor(input) {
+    this.columnMappedKeys = {};
+    this.inputPath = input;
+    this.dbClient;
+    console.log("this is csv directory from user:", this.inputPath);
+  }
+
+  csvToPg = async () => {
+    let linesCounter = 0;
```
Comment (asafMasa) [inline src/dataSource.js:55]: add logs for each phase, for example: 
"Start to read csv data"
"Handling row {n}"
"Insert row {n} to DB"
etc.

File: src/dataSource.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    this.dbClient = new DBProvider();
+    await this.constructHeaders(); // load actual column header name and map it by indexes
+    return new Promise((resolve, reject) => {
+      fs.createReadStream(this.inputPath)
+        .pipe(parse({ delimiter: ",", from_line: 2 }))
+        .on("error", reject)
+        .on("data", async (row) => {
+          linesCounter+=1;
+          if (row[this.columnMappedKeys['classification']] && CLASSIFICATION_MAPPING[row[this.columnMappedKeys['classification']].toLowerCase()]){
```
Comment (asafMasa) [inline src/dataSource.js]: why do you only check classification?
recordId for example is also a must - lets talk about required fields

File: src/dataSource.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+          const polygons = this.multiPolygon2Polygons(row[this.columnMappedKeys['wkt']])
+          let polygonCounter = 0;
+          for (const polygon of polygons){
+              polygonCounter+=1;
+              console.log(`Processing line number ${linesCounter} -- polygon ${polygonCounter}/${polygons.length}`)
+                const polygonParts = {
+                    recordId: row[this.columnMappedKeys['recordid']] ? row[this.columnMappedKeys['recordid']]: 'unknown',
```
Comment (asafMasa) [inline src/dataSource.js]: recordId - must be provided

File: src/dataSource.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                    productId: row[this.columnMappedKeys['productid']] ? row[this.columnMappedKeys['productid']]: 'unknown',
+                    productName: row[this.columnMappedKeys['productname']] ? row[this.columnMappedKeys['productname']]: 'unknown',
+                    productVersion: "1.0",
+                    productType: row[this.columnMappedKeys['producttype']] ? row[this.columnMappedKeys['producttype']]: 'orthophotoBest',
```
Comment (asafMasa) [inline src/dataSource.js]: Because recordId - must be provided, i prefer this value will also be provided

File: src/dataSource.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                    imageName: row[this.columnMappedKeys['imagename']] ? row[this.columnMappedKeys['imagename']]: undefined,
+                    sourceStartDateUtc: row[this.columnMappedKeys['sourcedateend']].replaceAll('/','-'),
```
Comment (asafMasa) [inline src/dataSource.js]: bug use "sourcedate**start**"

File: src/dataSource.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                    sourceEndDateUtc: row[this.columnMappedKeys['sourcedateend']].replaceAll('/','-'),
+                    minResolutionDegree: parseFloat(row[this.columnMappedKeys['maxresolutiondeg']]),
```
Comment (asafMasa) [inline src/dataSource.js]: bug use "**min**resolutiondeg"

File: src/dataSource.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                    maxResolutionDegree: parseFloat(row[this.columnMappedKeys['maxresolutiondeg']]),
+                    minResolutionMeter: parseFloat(row[this.columnMappedKeys['maxresolutionmeter']]),
```
Comment (asafMasa) [inline src/dataSource.js]: bug use "**min**resolutionmeter"

File: src/dataSource.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    this.dbClient = new DBProvider();
+    await this.constructHeaders(); // load actual column header name and map it by indexes
+    return new Promise((resolve, reject) => {
+      fs.createReadStream(this.inputPath)
+        .pipe(parse({ delimiter: ",", from_line: 2 }))
+        .on("error", reject)
+        .on("data", async (row) => {
+          linesCounter+=1;
+          if (row[this.columnMappedKeys['classification']] && CLASSIFICATION_MAPPING[row[this.columnMappedKeys['classification']].toLowerCase()]){
+          const polygons = this.multiPolygon2Polygons(row[this.columnMappedKeys['wkt']])
```
Comment (asafMasa) [inline src/dataSource.js]: throw error if wkt isn't provided

File: src/dataSource.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            console.error(`No classification provided for line ${linesCounter}, will skip this line`)
+          }
+        })
+        .on("end", () => {
+          resolve();
+        });
+    });
+  };
+
+  constructHeaders = async () => {
```
Comment (asafMasa) [inline src/dataSource.js]: is this method name explains what the method does?

File: src/dataSource.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+          }
+        })
+        .on("end", () => {
+          this.columnMappedKeys = { ...enumerableHeaders };
+          resolve();
+        });
+    });
+  };
+
+  printObjects = () => {
```
Comment (asafMasa) [inline src/dataSource.js]: is this method used? where?

File: src/dataSource.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        })
+        .on("end", () => {
+          resolve();
+        });
+    });
+  };
+
+  constructHeaders = async () => {
+    return new Promise((resolve, reject) => {
+      let enumerableHeaders = {};
```
Comment (asafMasa) [inline src/dataSource.js:153]: rename this variable name so it will be self explenatory

File: src/dataSource.js [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    });
+  };
+
+  printObjects = () => {
+    console.log(this.columnMappedKeys);
+  };
+
+
+  multiPolygon2Polygons = (wkt) => {
+    const geoJson = geoParse(wkt);
```
Comment (asafMasa) [inline src/dataSource.js:249]: check that it's realy multi polygon before trying to split it, if its type is polygon take it as is, multi-polygon split to polygons, other types throw an error

**Review by asafMasa — CHANGES_REQUESTED** [review]
Looking Good👍, please mind my remarks

File: src/dataSource.js (hunk 8 above)
Comment (ronenkapelian) [reply src/dataSource.js]: I was discuss with Shaziri, he will provide me only "max" and told me refernce the min the same as max

File: src/dataSource.js (hunk 9 above)
Comment (ronenkapelian) [reply src/dataSource.js]: I was discuss with Shaziri, he will provide me only "max" and told me refernce the min the same as max

File: src/dataSource.js (hunk 7 above)
Comment (ronenkapelian) [reply src/dataSource.js]: I was discuss with Shaziri, he will provide me only "end" and told me refernce the "start" the same as "end"

File: package.json (hunk 2 above)
Comment (ronenkapelian) [reply package.json]: done

File: src/dataSource.js (hunk 3 above)
Comment (ronenkapelian) [reply src/dataSource.js:55]: added

File: src/dataSource.js (hunk 10 above)
Comment (ronenkapelian) [reply src/dataSource.js]: done

File: src/dataSource.js (hunk 5 above)
Comment (ronenkapelian) [reply src/dataSource.js]: done

File: src/dataSource.js (hunk 14 above)
Comment (ronenkapelian) [reply src/dataSource.js:249]: done

File: src/dataSource.js (hunk 13 above)
Comment (ronenkapelian) [reply src/dataSource.js:153]: removed

File: src/dataSource.js (hunk 11 above)
Comment (ronenkapelian) [reply src/dataSource.js]: added remarks

File: src/dataSource.js [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+    await this.constructCsvHeaders(); // load actual column header name and map it by indexes
+    console.log(`--Loading CSV file to memory--`);
+    return new Promise((resolve, reject) => {
+      fs.createReadStream(this.inputPath)
+        .pipe(parse({ delimiter: ",", from_line: 2 }))
+        .on("error", reject)
+        .on("data", async (row) => {
+          linesCounter += 1;
+          const isValidRow = this.validateData(row); // checking mandatory fields
+          if (isValidRow) {
```
Comment (asafMasa) [inline src/dataSource.js]: you should change the oreder - if !valid - return

File: src/dataSource.js [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+        )}`
+      );
+    }
+
+    if (!row[this.columnMappedKeys["recordid"]]) {
+      isValid = false;
+      this.printValidationError('recordId');
+    }
+
+    if (!row[this.columnMappedKeys["wkt"]]) {
```
Comment (asafMasa) [inline src/dataSource.js:204]: you should validate that type is multipolygon or polygon

File: src/dataSource.js [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+                  row[this.columnMappedKeys["maxresolutionmeter"]]
+                ),
+                maxResolutionMeter: parseFloat(
+                  row[this.columnMappedKeys["maxresolutionmeter"]]
+                ),
+                minHorizontalAccuracyCe90: parseFloat(
+                  row[this.columnMappedKeys["minhorizontalaccuracyce90"]]
+                ),
+                sensors: row[this.columnMappedKeys["sensors"]],
+                region: row[this.columnMappedKeys["region"]]
```
Comment (asafMasa) [inline src/dataSource.js]: make sure sensors array is being recieved with required delimiter

File: src/dataSource.js [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+                  ? row[this.columnMappedKeys["region"]]
+                  : "unknown",
```
Comment (asafMasa) [inline src/dataSource.js]: make sure region array is being recieved with required delimiter

**Review by asafMasa — CHANGES_REQUESTED** [review]
👍 Good Job

File: src/dataSource.js (hunk 15 above)
Comment (ronenkapelian) [reply src/dataSource.js]: replaced and refactored

File: src/dataSource.js (hunk 16 above)
Comment (ronenkapelian) [reply src/dataSource.js:204]: added validation

File: src/dataSource.js (hunk 17 above)
Comment (ronenkapelian) [reply src/dataSource.js]: not on CLI scoop - will implemented on PolygonParts manager service on future

File: src/dataSource.js (hunk 18 above)
Comment (ronenkapelian) [reply src/dataSource.js]: not on CLI scoop - will implemented on PolygonParts manager service on future

File: src/dataSource.js [hunk 19]
```diff
@@ -193,32 +198,42 @@ class FileToDB {
...
+      this.printValidationError("recordId");
     }
 
     if (!row[this.columnMappedKeys["wkt"]]) {
       isValid = false;
-      this.printValidationError('wkt');
+      this.printValidationError("wkt");
+    } else {
+      const wkt = row[this.columnMappedKeys["wkt"]];
+      const geoType = wkt.split(" ")[0];
```
Comment (asafMasa) [inline src/dataSource.js:209]: 🤮

**Review by asafMasa — APPROVED** [review]

**Pull request merged**